codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/OperationContextParamsKnowledgeIndex.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.rules;
+
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.jr.stree.JrsString;
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.CodeBlock;
+import com.squareup.javapoet.MethodSpec;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.lang.model.element.Modifier;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.model.intermediate.OperationModel;
+import software.amazon.awssdk.codegen.model.rules.endpoints.ParameterModel;
+import software.amazon.awssdk.codegen.poet.PoetExtension;
+import software.amazon.awssdk.codegen.poet.waiters.JmesPathExpressionConverter;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.utils.CollectionUtils;
+
+/**
+ * Knowledge index to get access to operation context parameters.
+ */
+public final class OperationContextParamsKnowledgeIndex {
```

Comment:
Nit: `KnowledgeIndex` sounds fancy, but I don't know what it means😅. Maybe just `OperationContextParamsHelper`?

Reply:
It's not my favorite either, but there is prior art from the SRA implementation and I'd rather not introduce a new term right now. We could rename all the KnowledgeIndex instances in a separate PR.

Reply:
Ah I see, we have more `Helper` than `KnowledgeIndex` though ;) Anyway, it's codegen class, so it doesn't really matter